### ISSUE TITLE
Beat the `lds` task from scale

### ISSUE THREAD

**Issue by haznai** [haznai/peopen#3]
The Main Goal of the issue is to write a new marimo notebook based on `dspy_lds_pipeline.py`.

We try to ace the `lds` benchmark of the [scale paper](https://huggingface.co/datasets/rcds/swiss_leading_decision_summarization) as a baseline for the approach of solving the `peopen` problems in a `dspy` way.

**What needs to be done in this issue**
- [ ] Improve upon `dspy_lds_pipeline.py`
	- Errors from 2024-06-24
		- Fix the error related to missing class definition in DSPy
		- Investigate and resolve the `NoneType` attribute error
		- Address the `ZeroDivisionError` in the DSPy optimize**r**: 
	- look into todos of `dspy_lds_pipeline.py` code 
- [x] Create a more sophisticated text transformation pipeline
	- currently, only a single module is used with a chonkers of a signature. 
		- → specialize
        - collect more metrics (see table below)
        - repair and implement some misfunctioning metrics from other code files
- [ ] Test Models
	- Remote
		- `chatgpt4turbo`
		- `claude`
		- `sonnet`
	- llamafile
		- [Phi-3](https://github.com/microsoft/Phi-3CookBook/blob/main/md/01.Introduce/Phi3Family.md)
			- "Phi-3-Small is also more intensively trained on multilingual datasets compared to Phi-3-Mini.
		- LLama3
		- [ ] Look into new models from june 2024
- [ ] Make each experiment run seperate logs in phoenix 
	- send evaluation metrics to phoenix
	- save model in phoenix
- [ ] Parametrize
	- which model to use
	- the optimizer accepts many parameters, make it paremetrizable

```python
# almost all of these parametercs can be optimized further
optimizer = MIPROv2(
	prompt_model=model,
	task_model=model,
	metric=MetricWrapper,
	init_temperature=0.1,
	num_candidates=1,
)
kwargs = dict(num_threads=1, display_progress=True, display_table=0)
compiled_optimizer = optimizer.compile(
	student=SummarizationCoT(),
	trainset=test_dataset["train"][:20],
	num_batches=3,
	max_bootstrapped_demos=1,
	max_labeled_demos=1,
	eval_kwargs=kwargs,
)
```
- [ ] Use MIPROv2
	- once we have the pipeline in place, let's use the full dataset — we only pass a subset of the `examples` to the optimizer currently

**scale paper benchmark on `lds`**

![Image](https://github.com/haznai/peopen/assets/43835582/78e41870-0420-4c8f-a5e0-c128df845168)

**Comment by haznai** [comment]
ChatGPT is a benchmark leader for bertscore—consider this as your baseline benchmark to beat in this issue

**Comment by haznai** [comment]
- [x] use `mipro`
- [x] make a list of useful models to use, extract from obsidian, downloaded models and newer released models
- [x] extract todos from #1

**Comment by haznai** [comment]
changed the title and added a more comprehensive issue description. moving to in progress :)

**Comment by haznai** [comment]
### 2024-06-25

#### Summary
- [x] Renamed `dspy_lds_pipeline.py` to `first_running_pipeline_experiments.py`.
- [x] Started new notebook `second_running_pipeline_experiments.py` for issue #3.
  - [x] til: Python `@dataclass` and `field(init=false)`.
  - [x] til: `gh2md` allows turning Github issues to markdown files.
- [x] Learnings about Marimo.
  - [x] til: Run a notebook as an app with `marimo run`.
  - [x] til: Change the layout to a grid layout.
  - [x] til: Run the notebook simply as a Python notebook.
  - [x] til: About `graalpy`.
  - [x] til: Best practices in Marimo cells: pure, functional code.
  - [x] til: About `@functools.cache`, a built-in cache in Python.
- [x] Updated [#3](https://github.com/haznai/peopen/issues/3) with good next steps.
- [x] Started with a server command; next, need to get it running in a Marimo app that acts as a dashboard.
- [x] Approach in the notebook didn't work; starting outside the notebook with a `justfile` implementation.

#### Insights & Challenges
- til: Python `@dataclass` is highly convenient.
  - til: Python is advancing rapidly, raising curiosity if Mojo will have similar features.
- til: Using `gh2md` to convert Github issues to markdown is very handy.
  - Downloading for `anki-llm` started.
- til: Exciting new features of Marimo like `marimo run` and grid layout.
  - Marimo encourages pure, functional code in cells for better efficiency.
- til: The in-built `@functools.cache` in Python enhances performance.
- The notebook approach for the server command did not work as planned; attempting a `justfile` implementation could be more efficient.
  - Premature optimization led to some delays but was a valuable learning experience.

#### Open Questions & Next Steps
- [ ] Explore functional programming in Python: https://wickstrom.tech/2024-05-23-statically-typed-functional-programming-python-312.html
- [ ] Initiate `justfile` implementation of starting the servers outside the notebook for better structure.

**Comment by haznai** [comment]
# Date: 2024-07-22

## Summary
- [x] Attempted to start Phoenix and Llamafile servers, encountered issues.
- [x] Decided to pull a dependency for Llamafile to start it via Python.
- [x] Selected Overmind over Hivemind for its tmux feature and integration with iTerm2.
- [x] Implemented live REPL (scratchpad) in Marimo.
- [x] Confirmed the possibility to run Marimo notebooks as normal Python files.
- [x] Installed Overmind and learned to use `env -i` for a clean environment.
- [x] Successfully connected Overmind with logging, resolved `sudo` password issue.
- [x] Encountered and attempted to troubleshoot an error in the notebook during training.
- [x] Began evaluating `dspy.evaluate`, initial results promising.

#### Insights & Challenges
- til Troubles with server initialization, particularly with the dependencies.
- til Learned about `env -i` for launching from a clean environment.
- til Marimo now supports live REPL for more interactive development.
- Encountered an error: `Error getting source code: could not find class definition. Running without program aware proposer.`
- Evaluations in `Phoenix` are not being logged correctly, need to refine this.
- til Identified potential issues with `setuptools`, especially when including extra `eval` options.

#### Open Questions & Next Steps
- [ ] Rank next priorities.
- [x] Ensure Phoenix logs persist between sessions.
- [ ] Automate evaluation logging: non-verbose/silent option for `evaluate`.
- [ ] Ensure metrics are logged correctly.
  - [ ] Add `evaluations` column in Phoenix logs.
- [ ] Investigate persistent logging reliability.
- [x] Resolve the `setuptools` error when updating `pyproject.toml` with extra `eval` option.
  - [x] Reset `evals` installations if errors persist.
- [ ] Verify if the compiler optimized properly by checking saved `.json` files.
- [ ] Investigate hyperparameters for `compile` and `logging` servers.

**Comment by haznai** [comment]
**2024-07-23**

#### Summary (achievements, what was done)

- Fixed venv setup
- Made notebook wait a bit to avoid race conditions
- Changed optimizer because mipro wasn't working

#### Insights & Challenges (what was learned, what was hard, and what didn't work)

- til: Procfiles are awesome.
- Dependencies can be tricky to manage, avoid playing with them too much.
- Need more structured approach to save training run metadata (parameters, evaluation results).
- Ensuring the Phoenix logging setup is persistent can be challenging.

#### Open Questions & Next Steps (what is still open, what needs to be done next)

- [ ] check notebook for refactoring steps

**Comment by haznai** [comment]
### 2024-08-01

#### Summary (achievements, what was done)
- [x] Added new notes
  - Functional approach, `@dataclass` with all parameters
  - Save model dynamically and save parameters
  - Metrics implementation 
  - Split big `dspy` modules into smaller ones
- [x] Confirmed `de`, `it`, and `fr` are the only languages in the dataset
- [x] Running model with ~56.47 metric, with about 1.5 for one full run of 4 runs default param
- [x] Barebones Phoenix setup done, no attributes/metadata passing

#### Insights & Challenges (what was learned, what was hard, and what didn't work)
- `til`: `frozen=true` in Python dataclasses for immutability
- `til`: `fields` with `@dataclass` is powerful 
- `til`: PEP 703 discussions are insightful for future immutability paradigms
- `til`: Inheritance with `@dataclass` in Python can be problematic
- `til`: `Predict` step is the final step in `dspy`
- `til`: `-c` and `-gpu` command line options for `.llamafile
- `Phoenix setup` lacks eval logging and metadata passing
- `til`: A single `hyperparams` dataclass mapping 1:1 with the database would be better when doing the functional core

#### Open Questions & Next Steps (what is still open, what needs to be done next)
- [ ] Refactor single `hyperparams` dataclass
- [ ] Reevaluate all TODOs in the `third_refactored_notebook` for network optimization
- [ ] Save evaluation dynamically as well
- [ ] check todos of obsidian noteboo

#### Additional Notes
- running with phi-3-medium overnight, but I don't have much hopes
- Running model with the refactored setup using Phi-3-medium-128k Q4 model takes ~24GB of RAM and 92% of the gpu -> how close to the limit is this?
- Considering truncation and closing applications to free up RAM for faster processing
- Evaluating large context models and dataset handling for future test runs

**Comment by haznai** [comment]
### 2024-08-26

#### Summary
  - Tried to beat the benchmark.
  - Ran several benchmarks using different models and configurations.
	  - Explored `mlx` for speeding up inference.
	  - local llms are too slow for quick prototyping ...
  - Decided to drop `miprov2` for now due to error
  - Started training with GPT-4 mini, changing all modules to `predict`.
	  - Added a teacher network and adjusted hyperparameters.
	  - added a `max_tokens` parameter
		  - didn't commit, check the changes out and rethinkt hem

#### Insights & Challenges
- **til**: Learned about speed discrepancies in different versions of Llama models
	- mlx is usually faster, but can hang
- Discovered issues with token length exceeding maximum context length in models.
	- Realized something was wrong with the `max_tokens` parameter, causing requests to fail.

#### Open Questions & Next Steps
- [ ] Lower temperature settings.
- [ ] If all testing successful, do bootstrap fine-tuning for an efficient program.
- [ ] Find new metrics for evaluation, urgently needed.
	- [ ] Change metric to BLEU.
- [ ] Evaluate BERTScore on only `de` (German).
- [ ] Investigate why the model doesn't give examples in the prompt.
- [ ] use my fly.io credit

**Comment by haznai** [comment]
### 2024-09-03

#### Summary (achievements, what was done)
- Achieved goal of having `student.predictors()` work -> initializing everything in the module is of importance

#### Insights & challenges (what was learned, what was hard and what didn't work)
- `til`: Always initialize the `modules` in the final module `__init__` in `dspy`.

#### Open questions & next steps (what is still open, what needs to be done next)
- [ ] Get notes from scratchpad.
- [ ] Look into the findings of the saved model.
- [ ] Get @todos from code to see what to do next.
- [ ] Change metrics to BERT score.
- [ ] miprov2 run  overnight?

**Comment by haznai** [comment]
running a `bootstrapfewshotwithrandomsearch` with `bertscore` overnight ...

results look good, need to be analyzed and evaluated further :)
- [ ] evaluate `BootStrapFewShotWithRandomSearch` model `models/2024-09-03_second_training_run.json`

**Comment by haznai** [comment]
patched `grounded_proposer.py` and proposed a pull request upstream: https://github.com/stanfordnlp/dspy/pull/1453

**Comment by haznai** [comment]
had a run with `miprov2` and saved the results. 
after a while, the run stopped evaluating.
if this was because of a coding mistake or not i don't know.

- [ ] evaluate miprov2 run

**Comment by haznai** [comment]
running `BootStrapFewShotWithRandomSearch` again ...

this time with much much more data.

**edit**: my gpt4-o-mini budget ran out - thank god calls are cached


- [ ] look at the traces in phoenix and see the latest prompts/signatures used and copy them over (modify saved bootstrapfewshot model file maybe?)

**Comment by haznai** [comment]
### 2024-09-19 
#### peop-3
- set up a datagrip connection to phoenix db, the database was too complicated to query though
- changed `peop-3.py` to do evaluations as well
	- evaluated on ~100 samples, in german language, bootstrapfewshow from `models/2024-09-03_second_training_run.json` has following bert score: 0.715
	- miprov2 couldn't be evaluated

**Comment by haznai** [comment]
Considering this task to be done, the main idea of the ticket was to get an understanding of how to work with dspy on a known (and similiar) task. all the learnings and code are documented in this ticket
